### PR TITLE
refactor: ledger & anoncreds typing, part 1

### PR DESCRIPTION
**Pull request by mirgee**
Introduces a new crate, `anoncreds_types` (derived from `anoncreds-rs`) with ledger and anoncreds types.

Starts typing `base_anoncreds` and `base_ledger` traits, where each implementation has the responsibility of converting between the trait interface types and corresponding types of the library used for the implementation (`indy-vdr`, `anoncreds-rs`).

Explicating the information required by each implementation and formalizing the type conversions helps assuring correctness at compile time. Besides type safety when compared to passing strings, using custom types in the traits (as opposed to using types provided by a 3rd party library) also clearly separates dependencies of the implementation from the dependencies of the public interface.

Some of the code adopted from `anoncreds-rs` still requires a (dedicated) cleanup.

### CODE REVIEW

File: aries/aries_vcx/src/common/keys.rs [hunk 1]
```diff
@@ -2,14 +2,15 @@ use aries_vcx_core::{
...
+use did_parser::Did;
 use serde_json::Value;
 
 use crate::errors::error::prelude::*;
 
 pub async fn rotate_verkey_apply(
     wallet: &impl BaseWallet,
     indy_ledger_write: &impl IndyLedgerWrite,
-    did: &str,
+    did: &Did,
```
Comment (Patrik-Stas) [inline aries/aries_vcx/src/common/keys.rs:13]: 💌

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]

File: aries/misc/anoncreds_types/src/data_types/macros.rs
[the remark this answers is not in the document]
Comment (mirgee) [reply aries/misc/anoncreds_types/src/data_types/macros.rs:1]: Hell no, this is not staying 🙅‍♂️

File: aries/misc/anoncreds_types/src/data_types/mod.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+pub mod identifiers;
+#[cfg(feature = "ledger")]
+pub mod ledger;
+#[cfg(feature = "messages")]
+pub mod messages;
```
[the remark this answers is not in the document]
Comment (mirgee) [reply aries/misc/anoncreds_types/src/data_types/mod.rs:5]: Not unnecessary, it is intended for non-ledger types in `BaseAnoncreds`. Maybe messages is a wrong name.

File: aries/aries_vcx/src/common/primitives/credential_definition.rs [hunk 3]
```diff
@@ -239,15 +241,15 @@ impl CredentialDef {
 pub async fn generate_cred_def(
     wallet: &impl BaseWallet,
     anoncreds: &impl BaseAnonCreds,
-    issuer_did: &str,
-    schema_id: &str,
-    schema_json: &str,
+    issuer_did: &Did,
+    schema_id: &SchemaId,
+    schema_json: Schema,
```
[the remark this answers is not in the document]
Comment (mirgee) [reply aries/aries_vcx/src/common/primitives/credential_definition.rs:246]: Yes I know and agree

File: aries/misc/anoncreds_types/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+    // General errors
+    Input,
+    IOError,
+    InvalidState,
+    Unexpected,
+    // Credential/proof errors
+    CredentialRevoked,
+    InvalidUserRevocId,
+    ProofRejected,
+    RevocationRegistryFull,
```
[the remark this answers is not in the document]
Comment (mirgee) [reply aries/misc/anoncreds_types/src/error.rs:24]: The entire error side of this crate needs to be reworked a simplified

**Review by Patrik-Stas — APPROVED** [review]

**Pull request merged**